### PR TITLE
Build out ZQuery's test suite

### PR DESCRIPTION
**Pull request by jessepinho**
ZQuery only had very rudimentary tests. As it increases in complexity (due to, e.g., the features added in https://github.com/penumbra-zone/web/pull/1298), it needs tests to cover these complexities. This PR adds tests for many use cases, as well as splits out some of the code from `index.ts` to make it more readable.

### CODE REVIEW

File: packages/zquery/src/index.ts [hunk 1]
```diff
@@ -114,170 +95,17 @@ export function createZQuery<
...
-            try {
-              const data = await props.fetch(...args);
-              props.set(prevState => ({ ...prevState, data }));
-            } catch (error) {
-              props.set(prevState => ({ ...prevState, error }));
-            }
-          }
-        },
-      },
-    },
```
Comment (jessepinho) [inline packages/zquery/src/index.ts:281]: This code is all moved into `getSlice()`, unchanged

File: packages/zquery/src/index.ts [hunk 2]
```diff
@@ -114,170 +95,17 @@ export function createZQuery<
...
-          return {
-            data: zQuery.data,
-            loading: zQuery.loading,
-            error: zQuery.error,
-          };
-        }),
-      );
-
-      return returnValue;
-    },
```
Comment (jessepinho) [inline packages/zquery/src/index.ts:198]: This code (and the deleted code above it) is moved into `getUseHook()`, unchanged

File: packages/zquery/src/get-slice.ts
Comment (jessepinho) [inline packages/zquery/src/get-slice.ts:1]: The code in this file is simply extracted, unchanged, from `index.ts`, to make the various files a little more readable/narrowly scoped

File: packages/zquery/src/get-use-hook.ts
Comment (jessepinho) [inline packages/zquery/src/get-use-hook.ts:1]: The code in this file is simply extracted, unchanged, from index.ts, to make the various files a little more readable/narrowly scoped

File: packages/zquery/src/index.test.ts [hunk 3]
```diff
@@ -1,57 +1,222 @@
...
+import { MOCK_PUPPY_PHOTOS, PuppyPhoto, State } from './test/mock-state';
+import { StoreApi, UseBoundStore, create } from 'zustand';
+import { ZQuery } from './types';
+import { waitFor } from '@testing-library/react';
 
 describe('createZQuery()', () => {
-  type MockState = null;
-  let mockUseStore: UseStore<MockState>;
-  let mockState: MockState;
+  let zQuery: ZQuery<'puppyPhotos', PuppyPhoto[], []>;
```
Comment (jessepinho) [inline packages/zquery/src/index.test.ts:9]: I've just tweaked the setup for the tests, but the first few tests are the same until down where I've commented below.

File: packages/zquery/src/index.test.ts [hunk 4]
```diff
@@ -1,57 +1,222 @@
...
+          revalidate: expect.any(Function),
+          _zQueryInternal: {
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+            fetch: expect.any(Function),
+            referenceCount: 0,
+          },
+        });
+      });
+
+      describe('the reducers returned by the `stream()` function', () => {
```
Comment (jessepinho) [inline packages/zquery/src/index.test.ts:60]: The tests in this suite are all new, and they test the various `on*()` reducers that you can pass to `createZQuery()`.

File: packages/zquery/src/get-slice.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+              ...prevState,
+              loading: false,
+              ...(onEnd ? { data: onEnd(prevState.data) } : {}),
+            }));
+          }
+        } else {
+          props.set(prevState => ({
+            ...prevState,
+            loading: true,
+          }));
```
Comment (jessepinho) [inline packages/zquery/src/get-slice.ts:112]: One change I did make is setting `loading` to `true` then `false` before and after a unary request, which I'd neglected to do previously.

**Review by grod220 — APPROVED** [review]
LGTM

**Pull request merged**